### PR TITLE
Added antler-proj integration

### PR DESCRIPTION
**Pull request by jolly-fellow**
Some commands may not work because they are not implemented in antler-proj (like clean build) or don't work in antler-proj yet (like removing of dependencies).

### CODE REVIEW

File: src/dune/__main__.py [hunk 1]
```diff
@@ -267,8 +272,148 @@ def load_all_modules_from_dir(plugin_dir):
...
+                                        args.update_dep[4])  # tag/release number
+                if len(args.update_dep) >= 4:
+                    dune_sys.update_dep(args.update_dep[0],  # project path
+                                        args.update_dep[1],  # object name (app/lib)
+                                        args.update_dep[2],  # dependency name
+                                        args.update_dep[3])  # location of dep
+                if len(args.update_dep) >= 3:
+                    dune_sys.update_dep(args.update_dep[0],  # project path
+                                        args.update_dep[1],  # object name (app/lib)
+                                        args.update_dep[2])  # dependency name
```
[the remark this answers is not in the document]
Comment (jolly-fellow) [reply src/dune/__main__.py:393]: > Same comment as add_args above. Please simplify.

I understand that it looks ugly but it happens because of the chaotic syntax of the whole CLI interface of DUNE. It make sense to redesign it, make it less ambiguous, use hierarchical approach with commands and their options, define a subparser for each command and describe the arguments correctly. In this case the parser will be able to generate correct help, recognize optional and demanded arguments correctly and we will not need to parse the arguments manually and add to their description "(Optional)". Parser should understand it itself. Anyway it is impossible to implement parsing of many optional arguments without a creation of a subparser object or manual parsing of the arguments. So I prefer to do it in a harmonious well designed system instead of adding crutches into the existing system. I decided current approach is the less painful temporary solution.

File: src/dune/__main__.py [hunk 2]
```diff
@@ -267,8 +272,148 @@ def load_all_modules_from_dir(plugin_dir):
...
+                                     args.add_lib[4])  # linker opts
+                elif len(args.add_lib) >= 4:
+                    dune_sys.add_lib(args.add_lib[0],  # path
+                                     args.add_lib[1],  # lib name
+                                     args.add_lib[2],  # lang
+                                     args.add_lib[3])  # compiler opts
+                elif len(args.add_lib) >= 3:
+                    dune_sys.add_lib(args.add_lib[0],  # path
+                                     args.add_lib[1],  # lib name
+                                     args.add_lib[2])  # lang
```
Comment (ScottBailey) [inline src/dune/__main__.py:314]:
```suggestion
                if len(args.add_lib) < 3 or len(args.add_lib) > 5:
                    parser.exit_with_help_message("--add-lib: invalid number of arguments")
                dune_sys.add_app(*args.add_lib)
```

File: tests/test_antler.py
Comment (ScottBailey) [inline tests/test_antler.py:1]: This should be renamed `test_antler_proj.py` or similar as there will be other antler inclusions later I believe.

Additionally, needs to be disabled for now.

File: tests/test_antler.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def remove_existing():
+    """ Remove an existing `TEST_PROJECT_DIR` dir. """
+
+    if os.path.exists(TEST_PROJECT_DIR):
+        print("Removing a test directory if exists: ", TEST_PROJECT_DIR)
+        shutil.rmtree(TEST_PROJECT_DIR)
+
+
```
Comment (ScottBailey) [inline tests/test_antler.py:39]:
```suggestion
@pytest.mark.skip(reason="Missing antler-proj release, to be fixed in issue DUNE#124")
```

**Review by ScottBailey — CHANGES_REQUESTED** [review]
With pytest name change and skip, as well as the eventual fix (#134) for the missing antler-proj release, we should be able to close this. Thanks!

**Comment by jolly-fellow** [comment]
> With pytest name change and skip, as well as the eventual fix (#134) for the missing antler-proj release, we should be able to close this. Thanks!

Done

File: src/dune/dune.py [hunk 4]
```diff
@@ -470,6 +470,148 @@ def system_newaccount(self, name, creator=None, pub=None, private=None, addition
...
+    def remove_lib(self, path: str,        # project path
+                   lib_name: str):       # lib name
+
+        container_dir = self._docker.abs_host_path(path)
+        self._docker.execute_cmd(["antler-proj", "remove", container_dir, "lib", lib_name])
+
+    def build_project(self, path, clean_build: bool = False):
+        container_dir = self._docker.abs_host_path(path)
+        if clean_build:
+            self._docker.execute_cmd(["antler-proj", "build", container_dir, "--clean"])
```
Comment (mikelik) [inline src/dune/dune.py:601]: Does you antler-proj handle `--clean`?
I took antler-proj from`main` and it doesn't.
If your antler-proj doesn't handle `--clean` as well can you please create an issue in antler-proj?

File: tests/test_antler_proj.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+def remove_existing():
+    """ Remove an existing `TEST_PROJECT_DIR` dir. """
+
+    if os.path.exists(TEST_PROJECT_DIR):
+        print("Removing a test directory if exists: ", TEST_PROJECT_DIR)
+        shutil.rmtree(TEST_PROJECT_DIR)
+
+
+@pytest.mark.skip(reason="Skeeped until the release of antler-proj. "
```
Comment (mikelik) [inline tests/test_antler_proj.py:41]:
```suggestion
@pytest.mark.skip(reason="Skipped until the release of antler-proj. "
```

File: tests/test_antler_proj.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    subprocess.run([DUNE_EXE, "--remove-app", TEST_PROJECT_DIR, "test_app"], check=True)
+    res = subprocess.run([DUNE_EXE, "--debug", "--add-app", TEST_PROJECT_DIR, "test_app",  "c++",
+                          "compiler_opts", "linker_opts"], stdout=subprocess.PIPE)
+    assert b'already exists in project' not in res.stdout
+
+    subprocess.run([DUNE_EXE, "--remove-lib", TEST_PROJECT_DIR, "test_lib"], check=True)
+    res = subprocess.run([DUNE_EXE, "--debug", "--add-lib", TEST_PROJECT_DIR, "test_lib",  "c++",
+                          "compiler_opts", "linker_opts"], stdout=subprocess.PIPE)
+    assert b'already exists in project' not in res.stdout
+
```
Comment (mikelik) [inline tests/test_antler_proj.py:83]: could you please add test for validate and build? This could be for this empty project as well.

**Review by ScottBailey — APPROVED** [review]
Address anything Mikal requires, but LGTM.

File: src/dune/__main__.py [hunk 7]
```diff
@@ -267,8 +272,66 @@ def load_all_modules_from_dir(plugin_dir):
...
                 handle_version()
-                dune_sys.execute_interactive_cmd(['apt','list','leap'])
-                dune_sys.execute_interactive_cmd(['apt','list','cdt'])
+                dune_sys.execute_interactive_cmd(['apt', 'list', 'leap'])
+                dune_sys.execute_interactive_cmd(['apt', 'list', 'cdt'])
+
+# Integration with antler-proj begin ----------------------------------------------------------------------
+
+            elif args.create_project:
+                dune_sys.create_project(args.create_project[1], args.create_project[0])
```
[the remark this answers is not in the document]
Comment (jolly-fellow) [reply src/dune/__main__.py:281]: > should we allow third optional argument: version? `antler-proj init [OPTIONS] path project_name [version]`

I can't imagine a case when a user of DUNE will want to create a new project with version not 0.0.1. But well I will add it now.

File: src/dune/dune.py (hunk 4 above)
Comment (jolly-fellow) [reply src/dune/dune.py:601]: > Does you antler-proj handle `--clean`? I took antler-proj from`main` and it doesn't. If your antler-proj doesn't handle `--clean` as well can you please create an issue in antler-proj?

No it doesn't bit this feature is required in #124 so we agreed format of the command with @ScottBailey and I made a corresponding issue https://github.com/AntelopeIO/antler-proj/issues/18

File: tests/test_antler_proj.py (hunk 6 above)
Comment (jolly-fellow) [reply tests/test_antler_proj.py:83]: > could you please add test for validate and build? This could be for this empty project as well.

We agreed with @ScottBailey to postpone the tests development and skip this file from active tests because antler-proj in active development stage and the tests will get broken after each major update i.e. almost every day. I removed all broken tests from this file and skipped it from the active tests.

File: tests/test_antler_proj.py (hunk 5 above)
Comment (jolly-fellow) [reply tests/test_antler_proj.py:41]: Done

**Review by mikelik — APPROVED** [review]
